Create Argentinian holidays provider.

nedSaf opened the pull request.

nedSaf commented:
## Update
- Added a new provider.
- Configured all the holidays and set all the descriptions from Wikipedia.

## ScreenShots
(This is a system we use for work logging, we pull the holidays into the work calendar)
![2021-11-18 12 21 21](https://user-images.githubusercontent.com/7369740/142397449-efa45d63-1c99-4321-884e-b9992093fcdc.gif)
![2021-11-18 12 22 12](https://user-images.githubusercontent.com/7369740/142397462-7dc4f7d1-2532-435e-8d30-88ae91cb2630.gif)


## Next
- Add unit test for all the holidays in this provider. (Copying the tests from other providers)

## TB
### 3h

mariano-dagostino commented on src/Yasumi/Provider/Argentina.php:
```diff
@@ -0,0 +1,271 @@
+     */
+    public function initialize(): void
+    {
+        $this->timezone = 'America/Santiago';
```
@nedSaf `America/Argentina/Buenos_Aires`

nedSaf replied:
@mariano-dagostino Thanks :)

mariano-dagostino reviewed (changes requested):
@nedSaf Checked all the days. They look good. The timezone is incorrect. Santiago is located in Chile.

stelgenhof commented:
Thanks @nedSaf for your work! Nice to have a another country added

nedSaf commented:
## Update
- Added proper naming of the holidays also in English.
- Added tests for all the holidays.

## ScreenShots
#### Unit test result
![Screen Shot 2021-11-19 at 11 52 55](https://user-images.githubusercontent.com/7369740/142603529-608e83cd-627a-479d-8a59-ce7420162e26.png)

### Holiday names in English
![2021-11-19 11 59 27](https://user-images.githubusercontent.com/7369740/142603891-145d1ce4-e92f-4d81-8d29-dfc024fede8b.gif)

## Review
@stelgenhof This is ready for your review.

/cc @mariano-dagostino

nedSaf commented:
Fixing tests...

mariano-dagostino commented on tests/Argentina/RemembranceDayTest.php:
```diff
@@ -0,0 +1,88 @@
+     * @throws Exception
+     * @throws ReflectionException
+     */
+    public function testDiaDosFinadosAfter2006(): void
```
```suggestion
    public function testRemembranceDayAfter2006(): void
```

mariano-dagostino commented on tests/Argentina/RemembranceDayTest.php:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @throws ReflectionException
+     */
+    public function testDiaDosFinadosBefore2006(): void
```
```suggestion
    public function testRemembranceDayBefore2006(): void
```

mariano-dagostino reviewed (approved):
@nedSaf looks good. There are two methods that could be renamed in the test.

stelgenhof commented:
@nedSaf You may want to check your tests as some of them are failing.

nedSaf commented:
@stelgenhof I fixed all the tests, ready for your review :)

stelgenhof commented on src/Yasumi/Provider/Argentina.php:
```diff
@@ -0,0 +1,307 @@
+                'en' => 'Carnival Monday',
+                'es' => 'Lunes de Carnaval',
+              ],
+              $carnavalMonday->sub(new DateInterval('P48D')),
```
The sub method can return `false` on failure, whilst the constructor 3rd parameter can only takes a `DateTimeInterface` object. It would be better to handle this situation.

stelgenhof commented on src/Yasumi/Provider/Argentina.php:
```diff
@@ -0,0 +1,307 @@
+                'en' => 'Carnival Tuesday',
+                'es' => 'Martes de Carnaval',
+              ],
+              $carnavalTuesday->sub(new DateInterval('P47D')),
```
The sub method can return `false` on failure, whilst the constructor 3rd parameter can only takes a `DateTimeInterface` object. It would be better to handle this situation.

stelgenhof commented on src/Yasumi/Provider/Argentina.php:
```diff
@@ -0,0 +1,307 @@
+     * @throws UnknownLocaleException
+     * @throws \Exception
+     */
+    public function initialize(): void
```
This method is pretty fat. I recommend to extract the custom holidays logic into separate (private) methods to make the code more cleaner.

stelgenhof replied:
I would make separate methods for each holiday, making it even more cleaner.

nedSaf replied:
Done.

stelgenhof commented on src/Yasumi/Provider/Argentina.php:
```diff
@@ -0,0 +1,307 @@
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @author Nader Safadi <nader@gizra.com>
```
Please leave the original author name in this file comment. You can add your own name in the class comment.

stelgenhof commented on tests/Argentina/ArgentinaTest.php:
```diff
@@ -0,0 +1,85 @@
+     */
+    public function testOfficialHolidays(): void
+    {
+        $this->assertDefinedHolidays([
```
This asserts that these holidays all are defined from the start test year (1980), which is correct. However it doesn't assert that the respective holidays are not present before their establishment year.

nedSaf replied:
But each test file has a method called `testNotHoliday`, which does exactly that (I know that they run for sure because they failed and I had to fix them), this structure of testing I copied from other providers so I think for consistency we should keep it as is.

But if you think otherwise, please let me know how you would like me to change the structure of the files?

stelgenhof replied:
You're correct; I confused with some other edge cases I worked on a while ago (Sometimes I don't even remember my own code ;) )

stelgenhof replied:
Just ran your test again and it fails:
```
1) Yasumi\tests\Argentina\ArgentinaTest::testObservedHolidays
Failed asserting that an array has the key 'carnavalMonday'.
```
You can replicate this by setting the starting test year lower than the establishment year. The error might not advertise itself as the actual test year is randomized.

nedSaf replied:
@stelgenhof I'm not sure what are you asking me to do here.

Each test has its own `ESTABLISHMENT_YEAR` on which the whole test is based, this is what I found in the other providers.

> you can replicate this by setting the starting test year lower than the establishment year.

Yes, but it's set to a specific year on `setUp`, it will not fail unless you change the year.
The tests will fail if you change anything in the code, in this state, it will not fail.

<img width="920" alt="Screen Shot 2021-11-24 at 9 13 22" src="https://user-images.githubusercontent.com/7369740/143191929-582a29ef-a4e1-44e1-b63a-48bf88889ea0.png">
<img width="932" alt="Screen Shot 2021-11-24 at 9 13 39" src="https://user-images.githubusercontent.com/7369740/143191938-57ecc4fd-d128-4c52-bb41-96fb35261cb4.png">



In any case, please let me know exactly what you want me to do if you think I'm wrong, I'm not sure what I need to do.

stelgenhof replied:
See for example this: https://github.com/azuyalabs/yasumi/blob/develop/tests/Georgia/GeorgiaTest.php#L65..L67

Your tests will not fail as your initial test start year (1980) is after any of the establishment years. It possibly can fail if you instantiate Yasumi with a year *before* any of the establishment years. Agreeably users might not do that a lot, but still is possible in case you'd like to see a backdated list of holidays.

stelgenhof commented on tests/Argentina/ArgentinaTest.php:
```diff
@@ -0,0 +1,85 @@
+     */
+    public function testObservedHolidays(): void
+    {
+        $this->assertDefinedHolidays([
```
This asserts that these holidays all are defined from the start test year (1980), which is correct. However it doesn't assert that the respective holidays are not present before their establishment year.

nedSaf replied:
Same as above ^^

nedSaf commented:
@stelgenhof Fixed your comments, with one comment from me ^^

stelgenhof reviewed (commented):
Please check my last comments.

nedSaf commented:
Pushed a fix.

nedSaf commented:
@stelgenhof Hey, I did push a fix after your comment, https://github.com/azuyalabs/yasumi/pull/264/commits/233f9d404e5f4f5f71eb20575805b5647dd4d0d0

I think that will cover the case you mentioned before, if not, please let me know.

nedSaf commented:
@stelgenhof any updates?
